bids/reports/report.py

```diff
@@ -49,6 +49,74 @@ def __init__(self, layout, config=None):
 
         self.config = config
 
+    def generate_from_files(self, files):
+        r"""Generate a methods section from a list of files.
+
+        Parameters
+        ----------
+        files : list of BIDSImageFile objects
+            List of files from which to generate methods description.
+
+        Returns
+        -------
+        counter : :obj:`collections.Counter`
+            A dictionary of unique descriptions across subjects in the file list,
+            along with the number of times each pattern occurred. In cases
+            where all subjects underwent the same protocol, the most common
+            pattern is most likely the most complete. In cases where the
+            file list contains multiple protocols, each pattern will need to be
+            inspected manually.
+
+        Examples
+        --------
+        >>> from os.path import join
+        >>> from bids.layout import BIDSLayout
+        >>> from bids.reports import BIDSReport
+        >>> from bids.tests import get_test_data_path
+        >>> layout = BIDSLayout(join(get_test_data_path(), 'synthetic'))
+        >>> report = BIDSReport(layout)
+        >>> files = layout.get(session='01', extension=['nii.gz', 'nii'])
+        >>> counter = report.generate_from_files(files)
+        Number of patterns detected: 1
+        Remember to double-check everything and to replace <deg> with a degree symbol.
+
+        >>> counter.most_common()[0][0]  # doctest: +ELLIPSIS
+        'For session 01:\n\tMR data were...'
+        """
+        descriptions = []
+
+        subjs = sorted(list(set([f.get_entities().get('subject') for f in files])))
+        sessions = sorted(list(set([f.get_entities().get('session') for f in files])))
+        for sid in subjs:
+            subject_files = [f for f in files if f.get_entities().get('subject') == sid]
```

Comment:
In practice this will rarely if ever make much of a difference, but in principle, if given a very large set of files, this might become slow, because you're unnecessarily looping over `files` again for each subject and session. You could potentially construct two dictionaries of lists outside the loop, and then just loop over their `items()`.